File: public/methods/10_FORECAST_EVALUATION/QM002_ROLLING_VS_EXPANDING_WINDOWS/labs/r/qm002_hands_on.R
# Reader-facing QM002 lab: compare expanding and rolling estimation windows.
args <- commandArgs(trailingOnly = TRUE)
window <- 60L
if (length(args) >= 1L) window <- as.integer(args[1L])
if (is.na(window) || window < 20L) stop("window must be at least 20")

script_arg <- commandArgs(trailingOnly = FALSE)
file_arg <- sub("^--file=", "", script_arg[grepl("^--file=", script_arg)])
script_path <- normalizePath(file_arg[1L])
root <- normalizePath(file.path(dirname(script_path), "..", ".."))
data_path <- file.path(root, "figures", "source_data", "qm002_series.csv")

fit_ar1 <- function(y, start0, end0) {
  # start0 and end0 use Python-style zero-based boundaries; targets are start0+1 ... end0-1.
  idx_target <- (start0 + 2L):end0
  idx_lag <- (start0 + 1L):(end0 - 1L)
  X <- cbind(1, y[idx_lag])
  solve(crossprod(X), crossprod(X, y[idx_target]))
}

rmse <- function(actual, forecast) sqrt(mean((actual - forecast)^2))

d <- read.csv(data_path)
y <- d$actual
n <- length(y)
break_t <- 220L
min_train <- 80L

out <- data.frame()
for (t0 in min_train:(n - 1L)) {
  be <- fit_ar1(y, 0L, t0)
  start0 <- max(0L, t0 - window - 1L)
  br <- fit_ar1(y, start0, t0)
  lag <- y[t0]
  row <- data.frame(
    t=t0,
    actual=y[t0 + 1L],
    expanding=be[1] + be[2]*lag,
    rolling=br[1] + br[2]*lag
  )
  out <- rbind(out, row)
}

post <- out$t >= break_t
exp_rmse <- rmse(out$actual[post], out$expanding[post])
roll_rmse <- rmse(out$actual[post], out$rolling[post])
pct <- 100*(roll_rmse/exp_rmse - 1)

cat(sprintf("Rolling window: %d\n", window))
cat(sprintf("Post-break RMSE — expanding: %.3f\n", exp_rmse))
cat(sprintf("Post-break RMSE — rolling:   %.3f\n", roll_rmse))
cat(sprintf("Rolling vs expanding:        %+.1f%%\n", pct))
cat(sprintf("QM002_LAB_METRIC window=%d\n", window))
cat(sprintf("QM002_LAB_METRIC post_rmse_expanding=%.12g\n", exp_rmse))
cat(sprintf("QM002_LAB_METRIC post_rmse_rolling=%.12g\n", roll_rmse))
cat(sprintf("QM002_LAB_METRIC post_rolling_vs_expanding_pct=%.12g\n", pct))

File: public/methods/10_FORECAST_EVALUATION/QM002_ROLLING_VS_EXPANDING_WINDOWS/labs/python/qm002_hands_on.py
"""Reader-facing QM002 lab: compare expanding and rolling estimation windows."""

from __future__ import annotations
from pathlib import Path
import argparse
import numpy as np
import pandas as pd

ROOT = Path(__file__).resolve().parents[2]
DATA = ROOT / "figures" / "source_data" / "qm002_series.csv"
BREAK_T = 220
MIN_TRAIN = 80


def fit_ar1(y: np.ndarray, start: int, end_exclusive: int) -> np.ndarray:
    target = y[start + 1:end_exclusive]
    lag = y[start:end_exclusive - 1]
    x = np.column_stack([np.ones(target.size), lag])
    return np.linalg.lstsq(x, target, rcond=None)[0]


def rmse(actual: np.ndarray, forecast: np.ndarray) -> float:
    return float(np.sqrt(np.mean((actual - forecast) ** 2)))


def evaluate(window: int) -> dict[str, float]:
    y = pd.read_csv(DATA)["actual"].to_numpy(float)
    rows = []
    for t in range(MIN_TRAIN, len(y)):
        beta_e = fit_ar1(y, 0, t)
        start = max(0, t - window - 1)
        beta_r = fit_ar1(y, start, t)
        rows.append((t, y[t], beta_e[0] + beta_e[1] * y[t-1], beta_r[0] + beta_r[1] * y[t-1]))
    d = pd.DataFrame(rows, columns=["t","actual","expanding","rolling"])
    post = d["t"] >= BREAK_T
    early = (d["t"] >= BREAK_T) & (d["t"] < BREAK_T + window)
    out = {
        "window": float(window),
        "post_rmse_expanding": rmse(d.loc[post,"actual"].to_numpy(), d.loc[post,"expanding"].to_numpy()),
        "post_rmse_rolling": rmse(d.loc[post,"actual"].to_numpy(), d.loc[post,"rolling"].to_numpy()),
        "early_post_rmse_expanding": rmse(d.loc[early,"actual"].to_numpy(), d.loc[early,"expanding"].to_numpy()),
        "early_post_rmse_rolling": rmse(d.loc[early,"actual"].to_numpy(), d.loc[early,"rolling"].to_numpy()),
    }
    out["post_rolling_vs_expanding_pct"] = 100*(out["post_rmse_rolling"]/out["post_rmse_expanding"]-1)
    return out


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--window", type=int, default=60)
    args = parser.parse_args()
    if args.window < 20:
        raise SystemExit("--window must be at least 20 for this exercise.")
    out = evaluate(args.window)
    print(f"Rolling window: {args.window}")
    print(f"Post-break RMSE — expanding: {out['post_rmse_expanding']:.3f}")
    print(f"Post-break RMSE — rolling:   {out['post_rmse_rolling']:.3f}")
    print(f"Rolling vs expanding:        {out['post_rolling_vs_expanding_pct']:+.1f}%")
    for k,v in out.items():
        print(f"QM002_LAB_METRIC {k}={v:.12g}")


if __name__ == "__main__":
    main()
